labgrid/remote/client.py

```diff
@@ -1530,6 +1531,85 @@ async def export(self, place, target):
     def print_version(self):
         print(labgrid_version())
 
+    def adb(self):
+        place = self.get_acquired_place()
+        target = self._get_target(place)
+        name = self.args.name
+        adb_cmd = ["adb"]
+
+        from ..resource.adb import NetworkADBDevice, RemoteADBDevice
+
+        for resource in target.resources:
+            if name and resource.name != name:
+                continue
+            if isinstance(resource, NetworkADBDevice):
+                host, port = proxymanager.get_host_and_port(resource)
+                adb_cmd = ["adb", "-H", host, "-P", str(port), "-s", resource.serial]
+                break
+            elif isinstance(resource, RemoteADBDevice):
+                host, port = proxymanager.get_host_and_port(resource)
+                # ADB does not automatically remove a network device from its
+                # devices list when the connection is broken by the remote, so the
+                # adb connection may have gone "stale", resulting in adb blocking
+                # indefinitely when making calls to the device. To avoid this,
+                # always disconnect first.
+                subprocess.run(["adb", "disconnect", f"{host}:{str(port)}"], stderr=subprocess.DEVNULL, timeout=10)
+                subprocess.run(
+                    ["adb", "connect", f"{host}:{str(port)}"], stdout=subprocess.DEVNULL, timeout=10
+                )  # Connect adb client to TCP adb device
+                adb_cmd = ["adb", "-s", f"{host}:{str(port)}"]
+                break
+
+        adb_cmd += self.args.leftover
+        subprocess.run(adb_cmd)
+
+    def scrcpy(self):
+        place = self.get_acquired_place()
+        target = self._get_target(place)
+        name = self.args.name
+        scrcpy_cmd = ["scrcpy"]
+        env_var = os.environ.copy()
+
+        from ..resource.adb import NetworkADBDevice, RemoteADBDevice
+
+        for resource in target.resources:
+            if name and resource.name != name:
+                continue
+            if isinstance(resource, NetworkADBDevice):
+                host, adb_port = proxymanager.get_host_and_port(resource)
+                ip_addr = gethostbyname(host)
+                env_var["ADB_SERVER_SOCKET"] = f"tcp:{ip_addr}:{adb_port}"
+
+                scrcpy_cmd = [
+                    "scrcpy",
+                    "--port",
+                    "27183",
+                    "-s",
+                    resource.serial,
+                ]
+
+                # If a proxy is required, we need to setup a ssh port forward for the port
+                # (27183) scrcpy will use to send data along side the adb port
+                if resource.extra.get("proxy_required") or self.args.proxy:
+                    proxy = resource.extra.get("proxy")
+                    scrcpy_cmd.append(f"--tunnel-host={ip_addr}")
+                    scrcpy_cmd.append(f"--tunnel-port={sshmanager.request_forward(proxy, host, 27183)}")
```

Comment:
Is this always the same port? How will this interact with multiple devices connected over ADB to the same host?

Reply:
Good point, I had not tested multiple simultaneous screen sharing sessions. I have added some code to find a free port on the exporter and use that. It is based on the "get_free_port" code within labgrid that gets a free port on the client machine.

Is this worth moving into a more general location? Maybe within `SSHConnection`?

Reply:
`get_free_port` should already be in `util.helper`, so you should be able to import it directly.

Reply:
That returns a free port on the machine running the labgrid client, I require a free port on the exporter. The way scrcpy works is that it runs a server on the android device and creates a port forward between it and the ADB host (the exporter in this case)

Reply:
You might be able to handle that within the remote resource for the exporter, the `SerialPortExport` does something similar to choose a free port for ser2net.

Reply:
There is an additional complication to the `scrcpy` command. 

I already use `get_free_port` in the `ADBExport` to get a port for the ADB server, which is started by the `ADBExport`.

The `scrcpy` server is not setup at this time, it is started by the `scrcpy` client when it connects to the device. It first connects via ADB, transfers over the binaries, starts the server, then uses ADB to create a port forward between the device and the ADB host (the exporter). 

I am unable to know at the time of starting the exporter what ports will be available at the time `scrcpy` is launched, which could be weeks later.